### PR TITLE
Adds sendmsg and recvmsg hooks mainly for DNS resolving (#1776).

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #1776 

The C# mongo client tries to resolve DNS with [`sendmsg`](https://man.freebsd.org/cgi/man.cgi?query=sendmsg&apropos=0&sektion=2&manpath=FreeBSD+13.1-RELEASE+and+Ports&arch=default&format=html) and [`recvmsg`](https://man.freebsd.org/cgi/man.cgi?query=recvmsg&apropos=0&sektion=2&manpath=FreeBSD+13.1-RELEASE+and+Ports&arch=default&format=html) which we were not hooking.

Now both are hooked, and work similar PR #1520 , where they only handle traffic on port `53` (it's tailor made for DNS resolution).

If the destination is on port `53`, we create an interceptor socket, and connect to it, otherwise we just perform a fake connection (change mirrord socket state, not actual socket).

### CODE REVIEW

File: mirrord/layer/src/socket/hooks.rs [hunk 1]
```diff
@@ -370,6 +371,128 @@ pub(super) unsafe extern "C" fn send_to_detour(
...
     }
 }
+
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn recvmsg_detour(
+    sockfd: i32,
+    message_header: *mut libc::msghdr,
+    flags: c_int,
+) -> ssize_t {
+    // Returns `-1`, but calling the original function sets whatever `errno` for us.
```
Comment (aviramha) [inline mirrord/layer/src/socket/hooks.rs:381]: I think you can just call recvmsg on the socket, then just modify the msg headers.. no?

File: mirrord/layer/src/socket/hooks.rs [hunk 2]
```diff
@@ -370,6 +371,128 @@ pub(super) unsafe extern "C" fn send_to_detour(
...
+                .reduce(|acc, _| {
+                    let block = data_blocks.byte_add(offset);
+                    offset += (*data_blocks).iov_len;
+
+                    acc + (*block).iov_len
+                })
+                .unwrap_or_default();
+            // TODO(alex) [mid] 2023-08-08: Do we need this `total_length` check though? Real use
+            // example will tell us the answer.
+            let layout = Layout::array::<u8>(total_length).unwrap();
```
Comment (aviramha) [inline mirrord/layer/src/socket/hooks.rs:401]: You can use more high level API for interacting with the cmsg.
https://docs.rs/nix/0.26.2/nix/sys/socket/enum.ControlMessageOwned.html

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord/layer/src/socket/hooks.rs (hunk 2 above)
Comment (meowjesty) [reply mirrord/layer/src/socket/hooks.rs:401]: Nice, had no idea about this.

File: mirrord/layer/src/socket/hooks.rs [hunk 3]
```diff
@@ -370,6 +371,62 @@ pub(super) unsafe extern "C" fn send_to_detour(
...
+    flags: c_int,
+) -> ssize_t {
+    debug!("recvmsg_detour -> fd {sockfd:#?}");
+    let recvmsg_result = FN_RECVMSG(sockfd, message_header, flags);
+
+    if recvmsg_result == -1 && errno::errno() != errno::Errno(libc::EAGAIN) {
+        recvmsg_result
+    } else {
+        if errno::errno() == errno::Errno(libc::EAGAIN) {
+            debug!("we hit EAGAIN");
```
Comment (aviramha) [inline mirrord/layer/src/socket/hooks.rs:391]: trace or remove?

File: mirrord/layer/src/socket/ops.rs [hunk 4]
```diff
@@ -431,7 +433,7 @@ fn connect_to_local_address(
 /// that will be handled by `(Tcp|Udp)OutgoingHandler`, starting the request interception procedure.
 ///
 /// 3. `sockt.state` is `Bound`: part of the tcp mirror feature.
-#[tracing::instrument(level = "trace", ret, skip(raw_address))]
```
Comment (aviramha) [inline mirrord/layer/src/socket/ops.rs:434]: trace?

File: mirrord/layer/src/socket/ops.rs [hunk 5]
```diff
@@ -699,7 +701,7 @@ pub(super) fn dup<const SWITCH_MAP: bool>(fd: c_int, dup_fd: i32) -> Result<(),
 ///
 /// `-layer` sends a request to `-agent` asking for the `-agent`'s list of `addrinfo`s (remote call
 /// for the equivalent of this function).
-#[tracing::instrument(level = "trace", ret)]
```
Comment (aviramha) [inline mirrord/layer/src/socket/ops.rs:702]: trace?

**Review by aviramha — APPROVED** [review]
Nice, small nitpicks

File: mirrord/layer/src/socket/hooks.rs (hunk 2 above)
Comment (meowjesty) [reply mirrord/layer/src/socket/hooks.rs]: For now, it seems like we're safe to ignore this stuff.

File: mirrord/layer/src/socket/hooks.rs (hunk 3 above)
Comment (meowjesty) [reply mirrord/layer/src/socket/hooks.rs]: trace all the things!

**Pull request merged**